File: test/unit/clients.test.js
"use strict";
const test = require("node:test");
const assert = require("node:assert");
const { Spotify, toTrack: spTrack, toAlbum: spAlbum, quoteTerm } = require("../../lib/spotify");
const { Qobuz, toTrack: qzTrack, toAlbum: qzAlbum } = require("../../lib/qobuz");
const PKCE = require("../../lib/spotify-pkce");

/** A fetch that answers from a scripted list and records what it was asked. */
function fakeFetch(script) {
  const calls = [];
  const queue = script.slice();
  const fn = async (url, init) => {
    calls.push({ url: String(url), init: init || {} });
    const next = queue.length > 1 ? queue.shift() : queue[0];
    return {
      ok: next.status >= 200 && next.status < 300,
      status: next.status,
      headers: { get: (k) => (next.headers || {})[String(k).toLowerCase()] || null },
      text: async () => (typeof next.body === "string" ? next.body : JSON.stringify(next.body)),
    };
  };
  fn.calls = calls;
  return fn;
}

const liveSession = { clientId: "cid", accessToken: "tok", refreshToken: "ref",
                      expiresAt: Date.now() + 600000, userId: "me" };

// ------------------------------------------------------------------- PKCE

test("PKCE verifier and challenge are well formed", () => {
  const v = PKCE.createVerifier();
  assert.strictEqual(v.length, 64);
  assert.match(v, /^[A-Za-z0-9\-._~]+$/);
  const c = PKCE.challengeFor(v);
  assert.match(c, /^[A-Za-z0-9\-_]+$/, "base64url, no padding");
  assert.strictEqual(PKCE.challengeFor(v), c, "the challenge is a pure function");
  // The published test vector from RFC 7636 appendix B.
  assert.strictEqual(
    PKCE.challengeFor("dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"),
    "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM");
});

test("checkRedirectUri enforces Spotify's loopback rule", () => {
  assert.ok(PKCE.checkRedirectUri("https://example.com/cb").ok);
  assert.ok(PKCE.checkRedirectUri("http://127.0.0.1:3380/api/spotify/callback").ok);
  const localhost = PKCE.checkRedirectUri("http://localhost:3380/cb");
  assert.strictEqual(localhost.ok, false);
  assert.match(localhost.reason, /127\.0\.0\.1/);
  const lan = PKCE.checkRedirectUri("http://192.168.1.50:3380/cb");
  assert.strictEqual(lan.ok, false);
  assert.match(lan.reason, /paste-back/);
});

test("parseCallback accepts a URL, a query string or a bare code", () => {
  assert.strictEqual(PKCE.parseCallback("http://x/cb?code=abc&state=s1").code, "abc");
  assert.strictEqual(PKCE.parseCallback("code=abc&state=s1").state, "s1");
  assert.strictEqual(PKCE.parseCallback("abc").code, "abc");
  assert.strictEqual(PKCE.parseCallback("http://x/cb?error=access_denied").error,
    "access_denied");
  assert.strictEqual(PKCE.parseCallback("").code, "");
});

test("buildAuthorizeUrl refuses to build half a sign-in", () => {
  assert.throws(() => PKCE.buildAuthorizeUrl({ redirectUri: "u", challenge: "c" }),
    /Client ID/);
  const url = PKCE.buildAuthorizeUrl({ clientId: "cid", redirectUri: "http://127.0.0.1/cb",
    challenge: "ch", state: "st" });
  const p = new URL(url).searchParams;
  assert.strictEqual(p.get("code_challenge_method"), "S256");
  assert.strictEqual(p.get("response_type"), "code");
  assert.ok(p.get("scope").includes("user-library-modify"));
});

// ------------------------------------------------------ Spotify conversion

test("Spotify tracks convert, and non-tracks are refused", () => {
  const t = spTrack({ id: "1", name: "Song", duration_ms: 200000, type: "track",
    artists: [{ name: "A" }, { name: "B" }], album: { name: "Rec" },
    external_ids: { isrc: "GBAYE0601498" } });
  assert.deepStrictEqual(t, { id: "1", isrc: "GBAYE0601498", title: "Song",
    artists: ["A", "B"], album: "Rec", durationMs: 200000 });
  assert.strictEqual(spTrack({ id: null, name: "Local" }), null, "a local file has no id");
  assert.strictEqual(spTrack({ id: "e1", type: "episode" }), null, "a podcast is not a track");
  assert.strictEqual(spTrack(null), null);
});

test("Spotify albums convert, including the release year", () => {
  assert.deepStrictEqual(spAlbum({ id: "a", name: "Rumours", total_tracks: 11,
    artists: [{ name: "Fleetwood Mac" }], external_ids: { upc: "075992736121" },
    release_date: "1977-02-04" }),
    { id: "a", upc: "075992736121", title: "Rumours", artists: ["Fleetwood Mac"],
      trackCount: 11, year: 1977 });
});

test("quoteTerm strips quotes that would break the search phrase", () => {
  assert.strictEqual(quoteTerm('The "Real" Thing'), '"The  Real  Thing"');
});

// -------------------------------------------------------- Qobuz conversion

test("Qobuz durations convert from seconds to milliseconds", () => {
  const t = qzTrack({ id: 7, title: "Song", duration: 213, isrc: "GBAYE0601498",
    performer: { name: "A" }, album: { title: "Rec" } });
  assert.strictEqual(t.durationMs, 213000,
    "seconds must become milliseconds or nothing ever matches");
});

test("Qobuz recomposes title and version into one title", () => {
  assert.strictEqual(
    qzTrack({ id: 1, title: "Blue Monday", version: "2016 Remaster", duration: 100 }).title,
    "Blue Monday (2016 Remaster)");
  assert.strictEqual(qzTrack({ id: 1, title: "Blue Monday", duration: 100 }).title,
    "Blue Monday");
});

test("Qobuz album tracks inherit the album block they came from", () => {
  const album = qzAlbum({ id: "al", title: "Rec", artist: { name: "A" }, tracks_count: 2 });
  const t = qzTrack({ id: 1, title: "Song", duration: 100 }, album);
  assert.strictEqual(t.album, "Rec");
  assert.deepStrictEqual(t.artists, ["A"]);
});

test("Qobuz albums convert, with the year from any of its date fields", () => {
  assert.strictEqual(qzAlbum({ id: 1, title: "X", release_date_original: "1977-02-04" }).year,
    1977);
  assert.strictEqual(qzAlbum({ id: 1, title: "X" }).year, null);
});

// ---------------------------------------------------------- rate limiting

test("a Spotify 429 is obeyed with the delay Spotify asked for", async () => {
  const fetch = fakeFetch([
    { status: 429, headers: { "retry-after": "1" }, body: "" },
    { status: 200, body: { id: "me", display_name: "Me" } },
  ]);
  const waits = [];
  const sp = new Spotify(liveSession, { fetch, onRateLimit: (ms) => waits.push(ms) });
  const me = await sp.me();
  assert.strictEqual(me.id, "me");
  assert.strictEqual(waits.length, 1);
  assert.ok(waits[0] >= 1000 && waits[0] <= 2000, "waited about the second it was told to");
});

test("a Spotify 401 refreshes once, then gives up rather than looping", async () => {
  const fetch = fakeFetch([{ status: 401, body: { error: { message: "expired" } } }]);
  const sp = new Spotify(Object.assign({}, liveSession, { refreshToken: "" }),
    { fetch });
  await assert.rejects(() => sp.me(), /expired/);
});

// ----------------------------------------------------------- write batching

test("Spotify playlist writes batch at 100 and use track URIs", async () => {
  const fetch = fakeFetch([{ status: 200, body: {} }]);
  const sp = new Spotify(liveSession, { fetch });
  await sp.addToPlaylist("pl", Array.from({ length: 250 }, (_, i) => "t" + i));
  assert.strictEqual(fetch.calls.length, 3, "250 tracks is three requests, not 250");
  const first = JSON.parse(fetch.calls[0].init.body);
  assert.strictEqual(first.uris.length, 100);
  assert.strictEqual(first.uris[0], "spotify:track:t0");
});

test("Spotify saved-track writes batch at 50", async () => {
  const fetch = fakeFetch([{ status: 200, body: {} }]);
  const sp = new Spotify(liveSession, { fetch });
  await sp.saveTracks(Array.from({ length: 120 }, (_, i) => "t" + i));
  assert.strictEqual(fetch.calls.length, 3);
  assert.strictEqual(JSON.parse(fetch.calls[0].init.body).ids.length, 50);
});

test("following artists puts the ids in the query, where Spotify wants them", async () => {
  const fetch = fakeFetch([{ status: 200, body: {} }]);
  const sp = new Spotify(liveSession, { fetch });
  await sp.followArtists(["a1", "a2"]);
  assert.match(fetch.calls[0].url, /ids=a1%2Ca2/);
  assert.match(fetch.calls[0].url, /type=artist/);
});

test("Qobuz writes batch at 50 and go in the query string", async () => {
  const fetch = fakeFetch([{ status: 200, body: { status: "success" } }]);
  const qz = new Qobuz({ token: "t", userId: "1" }, { fetch });
  await qz.addToPlaylist("pl", Array.from({ length: 130 }, (_, i) => "t" + i));
  assert.strictEqual(fetch.calls.length, 3);
  assert.match(fetch.calls[0].url, /track_ids=/);
  assert.match(fetch.calls[0].url, /app_id=/);
});

test("empty write lists send no requests at all", async () => {
  const fetch = fakeFetch([{ status: 200, body: {} }]);
  const sp = new Spotify(liveSession, { fetch });
  await sp.saveTracks([]);
  await sp.saveAlbums([null, ""]);
  assert.strictEqual(fetch.calls.length, 0);
});

// ------------------------------------------------------------------ paging

test("Spotify paging stops at the reported total", async () => {
  const page = (n, total) => ({ status: 200,
    body: { total, items: Array.from({ length: n }, (_, i) => ({ id: "p" + i })) } });
  const fetch = fakeFetch([page(50, 60), page(10, 60), page(0, 60)]);
  const sp = new Spotify(liveSession, { fetch });
  const all = await sp.pageAll("/me/playlists", {}, null, 50);
  assert.strictEqual(all.length, 60);
  assert.strictEqual(fetch.calls.length, 2, "it stopped without a third empty request");
});

test("Qobuz paging stops on a short page", async () => {
  const fetch = fakeFetch([
    { status: 200, body: { tracks: { items: Array.from({ length: 2 }, (_, i) => (
      { id: i, title: "T" + i, duration: 100 })) } } },
  ]);
  const qz = new Qobuz({ token: "t" }, { fetch });
  const all = await qz.savedTracks();
  assert.strictEqual(all.length, 2);
  assert.strictEqual(fetch.calls.length, 1);
});

// --------------------------------------------------------------------------
// Barcode search. Spotify's album search returns SimplifiedAlbumObject with
// no external_ids, so a candidate never carries a barcode and the matcher's
// barcode tier could not fire. The `upc:` filter is the fix, and because the
// result still has no barcode on it, THE FILTER IS THE EVIDENCE — the code is
// stamped on so the tier can see it.

test("a Spotify barcode search uses the upc: filter and stamps the result", async () => {
  const http = fakeFetch([{ status: 200, body: { albums: { items: [
    { id: "sal", name: "Master of Puppets", total_tracks: 8,
      artists: [{ name: "Metallica" }] },   // note: NO external_ids, as Spotify sends
  ] } } }]);
  const sp = new Spotify(liveSession, { fetch: http });
  const albums = await sp.searchByUpc("075992736121");
  assert.match(http.calls[0].url, /q=upc%3A075992736121/);
  assert.match(http.calls[0].url, /type=album/);
  assert.strictEqual(albums.length, 1);
  assert.strictEqual(albums[0].upc, "075992736121",
    "the searched-for barcode is stamped on, or the matcher's tier cannot fire");
});

test("a barcode search that returns a crowd is NOT trusted as one", async () => {
  // A barcode identifies one release. If the filter hands back a pile, it is
  // not behaving like a filter, and stamping would be a confidently wrong
  // match — the exact failure this app refuses to make.
  const many = Array.from({ length: 6 }, (_, i) =>
    ({ id: "a" + i, name: "Something Else", total_tracks: 9, artists: [{ name: "Nope" }] }));
  const http = fakeFetch([{ status: 200, body: { albums: { items: many } } }]);
  const sp = new Spotify(liveSession, { fetch: http });
  const albums = await sp.searchByUpc("075992736121");
  assert.strictEqual(albums.length, 6);
  assert.ok(albums.every((a) => a.upc === ""),
    "nothing is stamped, so these are judged on title and artist like anything else");
});

test("an empty barcode makes no request at all", async () => {
  const http = fakeFetch([{ status: 200, body: {} }]);
  const sp = new Spotify(liveSession, { fetch: http });
  assert.deepStrictEqual(await sp.searchByUpc(""), []);
  assert.deepStrictEqual(await sp.searchByUpc(null), []);
  assert.strictEqual(http.calls.length, 0);
});

test("Qobuz searches the barcode as a plain query and needs no stamp", async () => {
  // Unlike Spotify's, Qobuz's album listings DO carry upc, so the matcher
  // compares the real codes itself.
  const http = fakeFetch([{ status: 200, body: { albums: { items: [
    { id: 7, title: "Master Of Puppets", upc: "075992736121", tracks_count: 8,
      artist: { name: "Metallica" } },
  ] } } }]);
  const qz = new Qobuz({ token: "t" }, { fetch: http });
  const albums = await qz.searchByUpc("075992736121");
  assert.match(http.calls[0].url, /query=075992736121/);
  assert.match(http.calls[0].url, /type=albums/);
  assert.strictEqual(albums[0].upc, "075992736121", "read from the response, not stamped");
});

// ------------------------------------------------- an album's track listing
//
// albumTracks was added to the interface and to both clients for the
// corroboration tier — the thing that decides a barcode-less album — and NOT
// ONE of these tests existed. Only the fake was tested, and a fake agrees with
// whatever you wrote. On a real library the tier then refused every album,
// which is exactly the failure mode this whole repository is shaped against:
// the corroboration silently could not read anything, so nothing matched.

test("a Qobuz album's tracks come from album/get, with no extra parameter", async () => {
  // album/get RETURNS THE TRACKS ON ITS OWN. MusicD-Remote — a Qobuz client
  // that works — asks for `{album_id}` and reads `tracks.items`, and passing
  // an `extra` value the API does not define is a request that can be refused
  // outright. A refused read here is not a visible error: safely() turns it
  // into "could not read the track listing" and every album is unmatched.
  const http = fakeFetch([{ status: 200, body: {
    id: "0060254776324", title: "Kind Of Blue", upc: "0060254776324",
    tracks_count: 5, artist: { name: "Miles Davis" },
    tracks: { items: [
      { id: 1, title: "So What", duration: 545, isrc: "USSM17700001" },
      { id: 2, title: "Freddie Freeloader", duration: 574 },
    ] },
  } }]);
  const qz = new Qobuz({ token: "t" }, { fetch: http });
  const tracks = await qz.albumTracks("0060254776324");

  assert.match(http.calls[0].url, /album\/get/);
  assert.match(http.calls[0].url, /album_id=0060254776324/);
  assert.ok(!/extra=/.test(http.calls[0].url),
    "album/get carries the tracks already; an undefined extra risks a refusal");

  assert.deepStrictEqual(tracks.map((t) => t.title), ["So What", "Freddie Freeloader"]);
  assert.strictEqual(tracks[0].durationMs, 545000, "Qobuz counts SECONDS");
  assert.strictEqual(tracks[0].isrc, "USSM17700001");
  assert.strictEqual(tracks[0].album, "Kind Of Blue",
    "the album block does not repeat per track, so it is threaded down");
});

test("a Qobuz album with no tracks block is an empty listing, not a throw", async () => {
  const http = fakeFetch([{ status: 200, body: { id: "x", title: "Nothing" } }]);
  const qz = new Qobuz({ token: "t" }, { fetch: http });
  assert.deepStrictEqual(await qz.albumTracks("x"), []);
});

test("a Spotify album's tracks are paged, and carry no ISRC", async () => {
  // SimplifiedTrackObject: no external_ids and no album block. That is fine
  // for what this is for — comparing TITLES against the other side's listing
  // — but a caller must not expect an ISRC from it.
  const page1 = { items: Array.from({ length: 50 }, (_, i) =>
    ({ id: "t" + i, name: "Track " + i, duration_ms: 200000, artists: [{ name: "Band" }] })),
    total: 52 };
  const page2 = { items: [
    { id: "t50", name: "Track 50", duration_ms: 1000, artists: [{ name: "Band" }] },
    { id: "t51", name: "Track 51", duration_ms: 1000, artists: [{ name: "Band" }] }],
    total: 52 };
  const http = fakeFetch([{ status: 200, body: page1 }, { status: 200, body: page2 }]);
  const sp = new Spotify(liveSession, { fetch: http });
  const tracks = await sp.albumTracks("sal");

  assert.match(http.calls[0].url, /albums\/sal\/tracks/);
  assert.strictEqual(tracks.length, 52, "a box set is more than one page");
  assert.strictEqual(tracks[0].title, "Track 0");
  assert.strictEqual(tracks[0].isrc, "", "a simplified track has none");
});

test("an album id is escaped rather than concatenated", async () => {
  // Qobuz ids are digits and Spotify's are base62, but they arrive from a
  // search response rather than from us, and a path built by concatenation is
  // one odd id away from requesting something else entirely.
  const http = fakeFetch([{ status: 200, body: { items: [], total: 0 } }]);
  const sp = new Spotify(liveSession, { fetch: http });
  await sp.albumTracks("a/b?c");
  assert.match(http.calls[0].url, /albums\/a%2Fb%3Fc\/tracks/);
});

File: android/core/src/test/kotlin/com/musicd/migrate/ClientsTest.kt
package com.musicd.migrate

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test
import org.json.JSONObject

/** The JavaScript client tests (test/unit/clients.test.js), translated. */
class ClientsTest {

    private fun liveSession() = SpotifySession(
        clientId = "cid", accessToken = "tok", refreshToken = "ref",
        expiresAt = System.currentTimeMillis() + 600_000, userId = "me")

    // ---------------------------------------------------------------- PKCE

    @Test fun `PKCE verifier and challenge are well formed`() {
        val v = Pkce.createVerifier()
        assertEquals(64, v.length)
        assertTrue(v.all { it.isLetterOrDigit() || it in "-._~" })
        val c = Pkce.challengeFor(v)
        assertTrue("base64url, no padding", c.all { it.isLetterOrDigit() || it in "-_" })
        assertEquals("the challenge is a pure function", c, Pkce.challengeFor(v))
        // The published test vector from RFC 7636 appendix B. If this passes,
        // the hashing and the base64url encoding are both right.
        assertEquals("E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM",
            Pkce.challengeFor("dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"))
    }

    @Test fun `checkRedirectUri enforces Spotify's loopback rule`() {
        assertTrue(Pkce.checkRedirectUri("https://example.com/cb").ok)
        assertTrue(Pkce.checkRedirectUri("http://127.0.0.1:3380/api/spotify/callback").ok)
        val localhost = Pkce.checkRedirectUri("http://localhost:3380/cb")
        assertEquals(false, localhost.ok)
        assertTrue(localhost.reason.contains("127.0.0.1"))
        val lan = Pkce.checkRedirectUri("http://192.168.1.50:3380/cb")
        assertEquals(false, lan.ok)
        assertTrue(lan.reason.contains("paste-back"))
    }

    @Test fun `parseCallback accepts a URL, a query string or a bare code`() {
        assertEquals("abc", Pkce.parseCallback("http://x/cb?code=abc&state=s1").code)
        assertEquals("s1", Pkce.parseCallback("code=abc&state=s1").state)
        assertEquals("abc", Pkce.parseCallback("abc").code)
        assertEquals("access_denied", Pkce.parseCallback("http://x/cb?error=access_denied").error)
        assertEquals("", Pkce.parseCallback("").code)
    }

    @Test fun `buildAuthorizeUrl refuses to build half a sign-in`() {
        try {
            Pkce.buildAuthorizeUrl("", "u", "c", "s")
            fail("a missing client id must not produce a URL")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("Client ID"))
        }
        val p = Pkce.parseQuery(
            Pkce.buildAuthorizeUrl("cid", "http://127.0.0.1/cb", "ch", "st")
                .substringAfter("?"))
        assertEquals("S256", p["code_challenge_method"])
        assertEquals("code", p["response_type"])
        assertTrue(p["scope"]!!.contains("user-library-modify"))
    }

    @Test fun `the Qobuz sign-in needs no registration and finds its code`() {
        val url = QobuzOAuth.buildAuthorizeUrl("http://127.0.0.1:3380/api/qobuz/oauth/callback")
        assertTrue(url.startsWith(QobuzOAuth.SIGNIN_URL))
        assertEquals("abc123",
            QobuzOAuth.extractCode("http://x/cb?code_autorisation=abc123"))
        assertEquals("abc123", QobuzOAuth.extractCode("abc123"))
        // A URL that carried no code is an error page or a cancelled sign-in.
        assertEquals("", QobuzOAuth.extractCode("http://x/cb?error=nope"))
    }

    // -------------------------------------------------- Spotify conversion

    @Test fun `Spotify tracks convert, and non-tracks are refused`() {
        val t = toTrack(JSONObject("""
            {"id":"1","name":"Song","duration_ms":200000,"type":"track",
             "artists":[{"name":"A"},{"name":"B"}],"album":{"name":"Rec"},
             "external_ids":{"isrc":"GBAYE0601498"}}"""))
        assertEquals(Track("1", "GBAYE0601498", "Song", listOf("A", "B"), "Rec", 200000), t)
        assertNull("a local file has no id",
            toTrack(JSONObject("""{"id":null,"name":"Local"}""")))
        assertNull("a podcast is not a track",
            toTrack(JSONObject("""{"id":"e1","type":"episode"}""")))
        assertNull(toTrack(null))
    }

    @Test fun `a JSON null is not read as the string null`() {
        // The bug this guards is invisible on the JVM if optString is used:
        // Android's org.json returns the LITERAL "null". A null isrc read that
        // way would be SEARCHED FOR on the other service.
        val t = toTrack(JSONObject("""
            {"id":"1","name":"Song","duration_ms":1000,"type":"track",
             "artists":[],"album":null,"external_ids":{"isrc":null}}"""))!!
        assertEquals("", t.isrc)
        assertEquals("", t.album)
    }

    @Test fun `Spotify albums convert, including the release year`() {
        assertEquals(Album("a", "075992736121", "Rumours", listOf("Fleetwood Mac"), 11, 1977),
            toAlbum(JSONObject("""
                {"id":"a","name":"Rumours","total_tracks":11,
                 "artists":[{"name":"Fleetwood Mac"}],
                 "external_ids":{"upc":"075992736121"},"release_date":"1977-02-04"}""")))
    }

    @Test fun `quoteTerm strips quotes that would break the search phrase`() {
        assertEquals("\"The  Real  Thing\"", quoteTerm("The \"Real\" Thing"))
    }

    // ---------------------------------------------------- Qobuz conversion

    @Test fun `Qobuz durations convert from seconds to milliseconds`() {
        val t = toQobuzTrack(JSONObject("""
            {"id":7,"title":"Song","duration":213,"isrc":"GBAYE0601498",
             "performer":{"name":"A"},"album":{"title":"Rec"}}"""), null)!!
        assertEquals("seconds must become milliseconds or nothing ever matches",
            213000L, t.durationMs)
    }

    @Test fun `Qobuz recomposes title and version into one title`() {
        assertEquals("Blue Monday (2016 Remaster)", toQobuzTrack(JSONObject(
            """{"id":1,"title":"Blue Monday","version":"2016 Remaster","duration":100}"""),
            null)!!.title)
        assertEquals("Blue Monday", toQobuzTrack(JSONObject(
            """{"id":1,"title":"Blue Monday","duration":100}"""), null)!!.title)
    }

    @Test fun `Qobuz album tracks inherit the album block they came from`() {
        val album = toQobuzAlbum(JSONObject(
            """{"id":"al","title":"Rec","artist":{"name":"A"},"tracks_count":2}"""))
        val t = toQobuzTrack(JSONObject("""{"id":1,"title":"Song","duration":100}"""), album)!!
        assertEquals("Rec", t.album)
        assertEquals(listOf("A"), t.artists)
    }

    @Test fun `Qobuz albums convert, with the year from any of its date fields`() {
        assertEquals(1977, toQobuzAlbum(JSONObject(
            """{"id":1,"title":"X","release_date_original":"1977-02-04"}"""))!!.year)
        assertNull(toQobuzAlbum(JSONObject("""{"id":1,"title":"X"}"""))!!.year)
    }

    // ------------------------------------------------------ rate limiting

    @Test fun `a Spotify 429 is obeyed with the delay Spotify asked for`() {
        val http = FakeHttp(listOf(
            FakeHttp.res(429, "", mapOf("retry-after" to "1")),
            FakeHttp.res(200, """{"id":"me","display_name":"Me"}""")))
        val waits = ArrayList<Long>()
        val sp = SpotifyClient(liveSession(), http, onRateLimit = { waits.add(it) },
            sleeper = { /* no real sleeping in a test */ })
        assertEquals("me", sp.me().id)
        assertEquals(1, waits.size)
        assertTrue("waited about the second it was told to",
            waits[0] in 1000..2000)
    }

    @Test fun `a Spotify 401 with no refresh token gives up rather than looping`() {
        val http = FakeHttp(listOf(
            FakeHttp.res(401, """{"error":{"message":"expired"}}""")))
        val sp = SpotifyClient(liveSession().copy(refreshToken = ""), http,
            sleeper = {})
        try {
            sp.me()
            fail("a dead sign-in must be reported")
        } catch (e: AuthError) {
            assertTrue(e.message!!.contains("expired"))
        }
    }

    @Test fun `a Qobuz 401 is an AuthError, not a generic failure`() {
        val qz = QobuzClient(QobuzSession(token = "t"),
            FakeHttp(listOf(FakeHttp.res(401, "nope"))), sleeper = {})
        try {
            qz.savedTracks()
            fail("a 401 must be an AuthError so the migration stops")
        } catch (e: AuthError) {
            assertTrue(e.message!!.contains("sign in again"))
        }
    }

    // ------------------------------------------------------ write batching

    @Test fun `Spotify playlist writes batch at 100 and use track URIs`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, "{}")))
        SpotifyClient(liveSession(), http, sleeper = {})
            .addToPlaylist("pl", (0 until 250).map { "t$it" })
        assertEquals("250 tracks is three requests, not 250", 3, http.calls.size)
        val first = JSONObject(http.calls[0].body!!)
        assertEquals(100, first.getJSONArray("uris").length())
        assertEquals("spotify:track:t0", first.getJSONArray("uris").getString(0))
    }

    @Test fun `Spotify saved-track writes batch at 50`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, "{}")))
        SpotifyClient(liveSession(), http, sleeper = {})
            .saveTracks((0 until 120).map { "t$it" })
        assertEquals(3, http.calls.size)
        assertEquals(50, JSONObject(http.calls[0].body!!).getJSONArray("ids").length())
    }

    @Test fun `following artists puts the ids in the query, where Spotify wants them`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, "{}")))
        SpotifyClient(liveSession(), http, sleeper = {}).followArtists(listOf("a1", "a2"))
        assertTrue(http.calls[0].url.contains("ids=a1%2Ca2"))
        assertTrue(http.calls[0].url.contains("type=artist"))
    }

    @Test fun `Qobuz writes batch at 50 and go in the query string`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, """{"status":"success"}""")))
        QobuzClient(QobuzSession(token = "t", userId = "1"), http, sleeper = {})
            .addToPlaylist("pl", (0 until 130).map { "t$it" })
        assertEquals(3, http.calls.size)
        assertTrue(http.calls[0].url.contains("track_ids="))
        assertTrue(http.calls[0].url.contains("app_id="))
    }

    @Test fun `empty write lists send no requests at all`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, "{}")))
        val sp = SpotifyClient(liveSession(), http, sleeper = {})
        sp.saveTracks(emptyList())
        sp.saveAlbums(listOf(""))
        assertEquals(0, http.calls.size)
    }

    // ---------------------------------------------------------------- paging

    @Test fun `Spotify paging stops at the reported total`() {
        fun page(n: Int, total: Int) = FakeHttp.res(200,
            """{"total":$total,"items":[""" +
                (0 until n).joinToString(",") { """{"id":"p$it"}""" } + "]}")
        val http = FakeHttp(listOf(page(50, 60), page(10, 60), page(0, 60)))
        val all = SpotifyClient(liveSession(), http, sleeper = {}).playlists()
        assertEquals(60, all.size)
        assertEquals("it stopped without a third empty request", 2, http.calls.size)
    }

    @Test fun `Qobuz paging stops on a short page`() {
        val http = FakeHttp(listOf(FakeHttp.res(200,
            """{"tracks":{"items":[{"id":0,"title":"T0","duration":100},
                                   {"id":1,"title":"T1","duration":100}]}}""")))
        val all = QobuzClient(QobuzSession(token = "t"), http, sleeper = {}).savedTracks()
        assertEquals(2, all.size)
        assertEquals(1, http.calls.size)
    }

    @Test fun `a Spotify playlist page skips local files and episodes, naming them`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, """
            {"total":3,"items":[
              {"is_local":true,"track":{"name":"bootleg.flac"}},
              {"is_local":false,"track":{"id":"e","type":"episode","name":"Some Show"}},
              {"is_local":false,"track":{"id":"t","type":"track","name":"Song",
                "duration_ms":1000,"artists":[{"name":"A"}],"album":{"name":"R"},
                "external_ids":{}}}
            ]}"""), FakeHttp.res(200, """{"total":3,"items":[]}""")))
        val tracks = SpotifyClient(liveSession(), http, sleeper = {}).playlistTracks("p")
        assertEquals(3, tracks.size)
        assertEquals("local file", tracks[0].skip)
        assertEquals("podcast episode", tracks[1].skip)
        assertNull(tracks[2].skip)
    }
    // ------------------------------------------------------- barcode search
    // Spotify's album search returns SimplifiedAlbumObject with no
    // external_ids, so a candidate never carries a barcode and Match's barcode
    // tier could not fire. The `upc:` filter is the fix, and because the result
    // still has no barcode on it, THE FILTER IS THE EVIDENCE — the code is
    // stamped on so the tier can see it. JS twin: test/unit/clients.test.js.

    @Test fun `a Spotify barcode search uses the upc filter and stamps the result`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, """
            {"albums":{"items":[
              {"id":"sal","name":"Master of Puppets","total_tracks":8,
               "artists":[{"name":"Metallica"}]}
            ]}}""")))   // note: NO external_ids, exactly as Spotify sends
        val albums = SpotifyClient(liveSession(), http, sleeper = {})
            .searchByUpc("075992736121")
        assertTrue(http.calls[0].url.contains("q=upc%3A075992736121"))
        assertTrue(http.calls[0].url.contains("type=album"))
        assertEquals(1, albums.size)
        assertEquals("the searched-for barcode must be stamped on, or the tier cannot fire",
            "075992736121", albums[0].upc)
    }

    @Test fun `a barcode search that returns a crowd is not trusted as one`() {
        // A barcode identifies one release. If the filter hands back a pile it
        // is not filtering, and stamping would be a confidently wrong match.
        val items = (0 until 6).joinToString(",") {
            """{"id":"a$it","name":"Something Else","total_tracks":9,
                "artists":[{"name":"Nope"}]}"""
        }
        val http = FakeHttp(listOf(FakeHttp.res(200, """{"albums":{"items":[$items]}}""")))
        val albums = SpotifyClient(liveSession(), http, sleeper = {})
            .searchByUpc("075992736121")
        assertEquals(6, albums.size)
        assertTrue("nothing is stamped, so these are judged on title and artist",
            albums.all { it.upc.isEmpty() })
    }

    @Test fun `an empty barcode makes no request at all`() {
        val http = FakeHttp(listOf(FakeHttp.res(200, "{}")))
        val sp = SpotifyClient(liveSession(), http, sleeper = {})
        assertEquals(emptyList<Album>(), sp.searchByUpc(""))
        assertEquals(emptyList<Album>(), sp.searchByUpc("   "))
        assertEquals(0, http.calls.size)
    }

    @Test fun `Qobuz searches the barcode as a plain query and needs no stamp`() {
        // Unlike Spotify's, Qobuz's album listings DO carry upc, so Match
        // compares the real codes itself.
        val http = FakeHttp(listOf(FakeHttp.res(200, """
            {"albums":{"items":[
              {"id":7,"title":"Master Of Puppets","upc":"075992736121",
               "tracks_count":8,"artist":{"name":"Metallica"}}
            ]}}""")))
        val albums = QobuzClient(QobuzSession(token = "t"), http, sleeper = {})
            .searchByUpc("075992736121")
        assertTrue(http.calls[0].url.contains("query=075992736121"))
        assertTrue(http.calls[0].url.contains("type=albums"))
        assertEquals("read from the response, not stamped",
            "075992736121", albums[0].upc)
    }

}
